mudslide/turbomole_model.py

```diff
@@ -0,0 +1,234 @@
+from scipy.special import erf
+import subprocess
+import turboparse
+import re
+import copy as cp
+
+import os
+import shutil
+
+from pathlib import Path
+
+from .electronics import ElectronicModel_ 
+
+from typing import Tuple, Any
+
+from .typing import ArrayLike, DtypeLike
+from .constants import eVtoHartree, amu_to_au
+from .periodic_table import masses
+
+def turbomole_is_installed():
+    # needs to have turbodir set
+    has_turbodir = "TURBODIR" in os.environ
+    # check that the scripts directory is available by testing for `sysname`
+    has_scripts = shutil.which("sysname") is not None
+    # check that the bin directory is available by testing for `sdg`
+    has_bin = shutil.which("sdg") is not None
+
+    return has_turbodir and has_scripts and has_bin
+
+class TMModel(ElectronicModel_):
+    def __init__(
+        self,
+        turbomole_dir: str,
+        states: ArrayLike,
+        sub_dir_stem:str = "traj",
+        representation: str = "adiabatic",
+        reference: Any = None,
+        turbomole_modules = {"gs_energy": "ridft", "gs_grads": "rdgrad", "es_grads": "egrad"}
+    ):
+        ElectronicModel_.__init__(self, representation=representation, reference=reference)
+
+        self.turbomole_dir = turbomole_dir
+        self.states = states
+        self.nstates_ = len(self.states)
+        
+        self.sub_dir_stem = sub_dir_stem
+        self.sub_dir_num = 0
+
+        self.turbomole_modules = turbomole_modules
+        self.turbomole_init()
+
+        assert turbomole_is_installed()
+        assert all([ shutil.which(x) is not None for x in self.turbomole_modules.values() ])
+
+    
+    def nstates(self):
+            return self.nstates_
+
+    def ndim(self):
+            return self.ndim_
+
+    def turbomole_init(self):
+        self.coord_path = Path(self.turbomole_dir)/"control"  
+        self.get_coords()
+
+    def get_coords(self):
+        coords = []
+        self.atom_order = []
+        self.mass = []
+
+        coord = subprocess.run(["sdg", "coord"], capture_output= True, text = True).stdout
+    
+        coord_list = coord.rstrip().split("\n")
+
+        for c in coord_list[1:]:
+            c_list = c.split()
+            self.atom_order.append(c_list[3])
+            coords.append([float(val) for val in c_list[:3]])
+            self.mass.append(3 * [masses[c_list[3]]])
+
+        self.ndim_ = 3 * len(coords)
+        self.X = np.array(coords, dtype=np.float64).reshape(self.ndim())
+        self.mass = np.array(self.mass, dtype=np.float64).reshape(self.ndim()) * amu_to_au
+
+
+    def update_coords(self, X):
+        self.coord_path = subprocess.run(["sdg", "-f", "coord"], capture_output=True, text=True).stdout.rstrip()
+        X = X.reshape((self.ndim() // 3, 3))
+        with open(self.coord_path, "r") as f:
+            lines = f.readlines()
+        regex = re.compile(r"\$coord\n|\$coord\s")
+        read_start = np.argwhere(np.array([bool(regex.match(line)) for line in lines]))[0][0]
```

Comment:
Simplify this. It's wasteful and hard to read as is. It will basically check every single line to see if it starts with `$coord`. Will it work if the line is ` $coord`? or `$coord ! a comment`?

Reply:
Ok I have tried this now !

Reply:
<img width="748" alt="Screen Shot 2022-04-14 at 5 54 41 PM" src="https://user-images.githubusercontent.com/46786940/163482884-f537fcf6-2aab-44ee-87b7-21fda076c42c.png">

OK I think this is good!

Reply:
Everything fails pretty catastrophically if the line starts with ` $coord` instead of `$coord` (just adding one space in front. Turbomole, on the other hand, works just fine in that case.

Reply:
<img width="547" alt="Screen Shot 2022-04-19 at 1 21 14 PM" src="https://user-images.githubusercontent.com/46786940/164060149-3ae312e4-15a2-441e-8c8e-7fd716bec4c5.png">


line 97 should fix that !